peer/cmd/request: Introduce request timeout option

nhoriguchi opened the pull request:
This pull request is separated from #127 for client request timeout part as recommended in [comment](https://github.com/hyperledger-labs/minbft/pull/127#pullrequestreview-322880648).

ynamiki commented on sample/peer/cmd/request.go:
```diff
@@ -65,9 +68,20 @@ type clientStack struct {
-	res := <-client.Request(req)
-	fmt.Println("Reply:", string(res))
+	var timeoutChan <-chan time.Time
+	timeout := viper.GetDuration("client.timeout")
```
Memo: Errors (giving "zzz" for the parameter, etc.) are ignored. I found some similar usages in the samples so I would assume this is our policy for error handling.

sergefdrv replied:
If the value cannot be parsed as duration (like "zzz"), `GetDuration` returns the default value of `time.Duration` (which is `time.Duration(0)`). If we'd like to detect this better, we could use `viper.Get` and convert the value using [`github.com/spf13/cast.ToDurationE`](https://godoc.org/github.com/spf13/cast#ToDurationE).

sergefdrv replied:
Even though this is just a sample application, this kind of error detection might help avoid surprising results if there is a simple typo in configuration. Maybe address in a separate issue? The same also applies to `sample/config` package.

nhoriguchi replied:
@ynamiki, @sergefdrv thank you for the comments, I fully agree with fixing error checking over the sample application in a separate PR.

ynamiki commented on sample/peer/cmd/request.go:
```diff
@@ -65,9 +68,20 @@ type clientStack struct {
+		fmt.Println("Reply:", string(res))
+	case <-timeoutChan:
+		fmt.Println("Client Request timer expired.")
+		os.Exit(1)
```
Memo: Output an error message to stdout (not stderr) and exit with 1 for an error case are also our conventions.

sergefdrv replied:
[Initially, `request` function would return an error](https://github.com/hyperledger-labs/minbft/blob/50d684f2aff06919cbcf40287937a6ef55140063/sample/peer/cmd/request.go#L70) that is propagated by `cobra` up the call stack, and finally reported [here](https://github.com/hyperledger-labs/minbft/blob/50d684f2aff06919cbcf40287937a6ef55140063/sample/peer/cmd/root.go#L48). Maybe we should restore that behavior (in a separate PR)? 

In that case, it seems not necessary to output the error here, because `sample/peer/cmd.Execute` will do that. We might also want to report the error to `stderr` rather than `stdout`.

ynamiki reviewed (approved):
I left some comments, but they are intended to clarify our coding conventions.

sergefdrv reviewed (approved):
I left some comments regarding error handling which we could solve as separate issues. This PR itself looks good and ready to merge. Thanks!

sergefdrv commented:
@Naoya-Horiguchi CircleCI seems to have stucked again... Would you mind amending the last commit and force-pushing to trigger the CI? Then this MR can be merged.

Commit:
peer/cmd/request: Introduce request timeout option

MinBFT never fails in theory as long as we meet the requirement of
"<= f faulty nodes", but unfortunately there's no bug-free softwares,
so it's better to prepare for such potential issues to avoid infinite wait.
This patch suggests to add timeout to the sample application.
Note that if timeout value is not specified, sample application waits
forever as it does now.

Signed-off-by: Naoya Horiguchi <n-horiguchi@ah.jp.nec.com>

nhoriguchi commented:
I did git-force-push and now CI seems to work well.
